In [1]:
import pandas as pd
import numpy as np
from scipy import stats
from sqlalchemy import create_engine

In [ ]:
engine = create_engine("mysql+pymysql://user:password@localhost:3306/schema")
transactions_df = pd.read_sql("SELECT * FROM transactions_clean", engine)
customers_df = pd.read_sql("SELECT * FROM customers", engine)
markets_df = pd.read_sql("SELECT * FROM markets_clean", engine)

In [45]:
df = transactions_df.merge(customers_df, how='inner', on='customer_code')
markets_df = markets_df.rename(columns={'markets_code':'market_code'})
df = df.merge(markets_df, how='inner', on='market_code')

In [46]:
df.dtypes
df['order_date'] = pd.to_datetime(df['order_date'])
df['order_year'] = df['order_date'].dt.year

**03 identified a small set of named accounts driving the decline. Does that concentration also have a geographic signature? And independent of those accounts where is genuine diversification actually happening?**

Findings in SQL:
- Revenue concentrated in North, decline in 2018-2019
    - Significant drop in Delhi NCR, Ahmedabad, Kanpur
- In Central, all markets drop except Bhopal(Mark007)
- 3/5 Markets in South show growth, 1 market shut down (0 revenue in 2019)
- Central : Decrease in order_volume, avg_order_val, and avg_basket_size
- North & South : Decrease in order_volume, increase in avg_order_val and avg_basket_size

In [47]:
zone_stats = df.groupby('zone').agg(
    revenue = ('sales_amount','sum'),
    order_volume = ('zone','count'),
    avg_order_value = ('sales_amount','mean'),
    avg_basket_size = ('sales_qty', 'mean')
)
zone_stats['revenue_share%'] = ((zone_stats['revenue']/df['sales_amount'].sum())*100).round(1)
zone_stats.sort_values(by='revenue_share%', ascending=False)

,revenue,order_volume,avg_order_value,avg_basket_size,revenue_share%
zone,,,,,
North,677091990.0,68216,9925.706433,18.667688,68.6
Central,263861014.0,72346,3647.209438,10.485873,26.7
South,45744764.0,8110,5640.538101,49.271023,4.6


- North holds 68.6% of revenue share, Central - 26.7%

In [48]:
def compute_zone_stats(year, df):
    mask = (df['order_year'] == year)
    stats = df[mask].groupby('zone').agg(
        revenue = ('sales_amount','sum'),
        order_volume = ('zone','count'),
        avg_order_value = ('sales_amount','mean'),
        avg_basket_size = ('sales_qty', 'mean')
    )
    return stats
zone_stats2018 = compute_zone_stats(2018,df)
zone_stats2019 = compute_zone_stats(2019,df)

In [49]:
cols = ['revenue_change%', 'order_vol_change%', 'avg_order_val_change', 'avg_basket_size_change']
zone_change_18_19 = pd.DataFrame(columns=cols)
zone_change_18_19['revenue_change%'] = ((zone_stats2019['revenue'] - zone_stats2018['revenue'])/zone_stats2018['revenue'])*100
zone_change_18_19['order_vol_change%'] = ((zone_stats2019['order_volume'] - zone_stats2018['order_volume'])/zone_stats2018['order_volume'])*100
zone_change_18_19['avg_order_val_change'] = zone_stats2019['avg_order_value'] - zone_stats2018['avg_order_value']
zone_change_18_19['avg_basket_size_change'] = zone_stats2019['avg_basket_size'] - zone_stats2018['avg_basket_size']
zone_change_18_19

,revenue_change%,order_vol_change%,avg_order_val_change,avg_basket_size_change
zone,,,,
Central,-11.627614,-1.318602,-408.424090,-1.805270
North,-21.543058,-26.659768,669.687722,1.928782
South,-17.838337,-21.138704,238.403579,12.539211


- 2018-2019: Revenue is declined across all zones, North - 21.5% decline, South - 17.8%.

- Dive into North Zone:
- T-testing on order value and basket size to define significance of North's growth (it has highest revenue drop)
- H0 -> The 2018 and 2019 metrics are similar
- p < 0.05 to reject the null hypothesis

In [50]:
north_orders_2018 = df.loc[(df['order_year'] == 2018) & (df['zone'] == 'North')]
north_orders_2018 = df.loc[ (df['order_year'] == 2018) & (df['zone'] == 'North')]
print(north_orders_2018['sales_amount'].std())
print(north_orders_2019['sales_amount'].std())
print(north_orders_2018['sales_qty'].std())
print(north_orders_2019['sales_qty'].std())

41558.140688973326
35819.83523079136
93.93392984566115
77.09002985074505


variances differ meaningfully between years, so Welch's t-test is appropriate

In [51]:
stats.ttest_ind(north_orders_2018['sales_amount'], north_orders_2019['sales_amount'], equal_var=False)

TtestResult(statistic=np.float64(-1.9659035965797231), pvalue=np.float64(0.04931520698018623), df=np.float64(50598.236308114334))

In [52]:
stats.ttest_ind(north_orders_2018['sales_qty'], north_orders_2019['sales_qty'], equal_var=False)

TtestResult(statistic=np.float64(-2.566020962859691), pvalue=np.float64(0.010290075261507983), df=np.float64(51279.027380310334))

North:
- Order Value: pvalue = 0.0493, suggesting a mild growth.
- Basket Size: pvalue= 0.01, strong Basket Size growth.  
Evidence: The decline is volume-driven, not value-driven.

In [53]:
def compute_markets_stats(zone,df):
    mask = (df['zone'] == zone)
    stats = df[mask].groupby(['markets_name']).agg(
        revenue = ('sales_amount','sum'),
        order_volume = ('market_code','count'),
        avg_order_value = ('sales_amount','mean'),
        avg_basket_size = ('sales_qty', 'mean')
    )
    return stats
north_markets_stats = compute_markets_stats('North',df)
mask_zone = df['zone'] == 'North'
north_markets_stats['revenue_share%'] = (north_markets_stats['revenue']/df.loc[mask_zone,'sales_amount'].sum())*100
north_markets_stats

,revenue,order_volume,avg_order_value,avg_basket_size,revenue_share%
markets_name,,,,,
Ahmedabad,132526737.0,20110,6590.091348,10.298558,19.572929
Delhi NCR,520853134.0,44384,11735.155326,22.305132,76.925018
Kanpur,13583923.0,2813,4828.980803,5.916815,2.006215
Lucknow,3094007.0,104,29750.067308,356.653846,0.456955
Patna,4428393.0,402,11015.902985,13.694030,0.654031
Surat,2605796.0,403,6465.995037,42.429280,0.384851


- 76% of North's revenue is concentrated in a single market *Delhi NCR*
- Lucknow, Patna, and Surat have very low order volume, and contribute 1Cr to the revenue combined.

In [54]:
rev_2018 = df.loc[(mask_zone)&(df['order_year'] == 2018)].groupby('markets_name')['sales_amount'].sum()
rev_2019 = df.loc[(mask_zone)&(df['order_year'] == 2019)].groupby('markets_name')['sales_amount'].sum()
north_markets_stats['revenue_change%_18_19'] = ((rev_2019 - rev_2018)*100)/rev_2018
north_markets_stats

,revenue,order_volume,avg_order_value,avg_basket_size,revenue_share%,revenue_change%_18_19
markets_name,,,,,,
Ahmedabad,132526737.0,20110,6590.091348,10.298558,19.572929,-15.946198
Delhi NCR,520853134.0,44384,11735.155326,22.305132,76.925018,-22.375597
Kanpur,13583923.0,2813,4828.980803,5.916815,2.006215,-25.113888
Lucknow,3094007.0,104,29750.067308,356.653846,0.456955,-30.992459
Patna,4428393.0,402,11015.902985,13.694030,0.654031,-40.435160
Surat,2605796.0,403,6465.995037,42.429280,0.384851,-60.358000


All the markets face revenue decline 2018-2019:
- Delhi NCR - ~22.3%, Kanpur - ~25%, Ahmedabad - ~16%
- Rest of the markets have very low order volume so not totally reliable for overall trend

In [55]:
(north_markets_stats.loc['Delhi NCR', 'revenue'] / df['sales_amount'].sum())*100

np.float64(52.7875050387263)

- Delhi NCR alone holds ~52.7% share of the overall revenue.

In [56]:
df[mask_zone].groupby(['markets_name'])['custmer_name'].nunique()

markets_name
Ahmedabad    14
Delhi NCR     5
Kanpur        7
Lucknow       1
Patna         1
Surat         1
Name: custmer_name, dtype: int64

- Lucknow, Patna, and Surat only have a single customer each. Hence, the low order volume.

In [57]:
customer_segment_df = pd.read_csv(r'E:\Projects\DataAnalysis\Git Repos\Revenue-Decline-Root-Cause-Analysis\data\rfm_segmentation.csv')
segment_df = customer_segment_df[['custmer_name', 'Segment']]

- rfm_segmentation.csv contains customer segments, fm Scores, change in F & M over 2018-2019 for each customer. The Segments are labels based on change.

In [58]:
north_cust_w_segment = df[mask].merge(segment_df, how='left',on='custmer_name')
print(north_cust_w_segment[['customer_code', 'custmer_name', 
    'market_code', 'markets_name',
    'Segment']].groupby(['markets_name', 'custmer_name'])['Segment'].unique())

markets_name  custmer_name           
Ahmedabad     Atlas Stores                      [Loyal - Stable]
              Control                    [Loyal - Deteriorating]
              Electricalsbea Stores         [Low Value - Stable]
              Excel Stores                            [Champion]
              Flawless Stores             [Low Value - Churning]
              Forward Stores                [Mid Tier - At Risk]
              Integration Stores            [Mid Tier - At Risk]
              Novus                         [Low Value - Stable]
              Propel                      [Frequent - Low Value]
              Relief                        [Mid Tier - Growing]
              Sage                        [Low Value - Churning]
              Sound                       [Frequent - Low Value]
              Surface Stores                 [Mid Tier - Stable]
              Unity Stores                  [Low Value - Stable]
Delhi NCR     Electricalsara Stores       [High Valu

- 4/5 Delhi NCR's customers are *"High Value - At Risk"*.
    - The 22% decline is concentrated among customers whose RFM profile shifted to *High Value - At Risk* (2018-2019)
- Kanpur's customers are almost all either Low Value or deteriorating  
- Ahmedabad has 5/14 stable customers, 1 *champion*, and 1 *Mid Tier - growing*  
    - While these customers backed, due to the rest of the customers Ahmedabad faced 16% decline (2018-2019)

Dive into South Zone, to test growth

In [59]:
south_markets_stats = compute_markets_stats('South',df)
mask_zone = (df['zone'] == 'South')
south_markets_stats['revenue_share%'] = (south_markets_stats['revenue']/df.loc[mask_zone,'sales_amount'].sum())*100
south_markets_stats

,revenue,order_volume,avg_order_value,avg_basket_size,revenue_share%
markets_name,,,,,
Bengaluru,373115.0,14,26651.071429,29.500000,0.815645
Bhubaneshwar,893857.0,112,7980.866071,133.741071,1.954009
Chennai,18227503.0,1030,17696.604854,49.344660,39.846097
Hyderabad,7436823.0,2034,3656.255162,38.293510,16.257211
Kochi,18813466.0,4920,3823.875203,51.927236,41.127037


In [60]:
rev_2018 = df.loc[(mask_zone)&(df['order_year'] == 2018)].groupby('markets_name')['sales_amount'].sum()
rev_2019 = df.loc[(mask_zone)&(df['order_year'] == 2019)].groupby('markets_name')['sales_amount'].sum()
south_markets_stats['revenue_change%_18_19'] = ((rev_2019 - rev_2018)*100)/rev_2018
south_markets_stats

,revenue,order_volume,avg_order_value,avg_basket_size,revenue_share%,revenue_change%_18_19
markets_name,,,,,,
Bengaluru,373115.0,14,26651.071429,29.500000,0.815645,NaN
Bhubaneshwar,893857.0,112,7980.866071,133.741071,1.954009,51.538882
Chennai,18227503.0,1030,17696.604854,49.344660,39.846097,-42.878083
Hyderabad,7436823.0,2034,3656.255162,38.293510,16.257211,14.401843
Kochi,18813466.0,4920,3823.875203,51.927236,41.127037,5.581353


2018-2019:
- Bhubaneshwar only has 112 orders, the growth rate is not totally reliable.
- Hyderabad with 2034 orders and Kochi with 4920 orders show growth of ~14% and ~5.5%
- Chennai sees ~42.8 decline
- Bengaluru doesn't show up in this comparison. It had revenue in 2018 and effectively none in 2019, a market that went to zero

In [61]:
df[mask_zone].groupby(['markets_name'])['custmer_name'].nunique()

markets_name
Bengaluru       6
Bhubaneshwar    1
Chennai         1
Hyderabad       1
Kochi           4
Name: custmer_name, dtype: int64

- Bengaluru had the most customers. 
    - Probably one-time purchases
- Kochi has 4 customers most out of all the growing markets.
- All other markets have 1 customer each.

In [62]:
south_cust_w_segment = df[mask_zone].merge(segment_df, how='left',on='custmer_name')
print(south_cust_w_segment[['customer_code', 'custmer_name', 
    'market_code', 'markets_name',
    'Segment']].groupby(['markets_name', 'custmer_name'])['Segment'].unique())

markets_name  custmer_name         
Bengaluru     Electricalsara Stores     [High Value - At Risk]
              Epic Stores                     [Loyal - Stable]
              Excel Stores                          [Champion]
              Flawless Stores           [Low Value - Churning]
              Nomad Stores             [Loyal - Deteriorating]
              Sound                     [Frequent - Low Value]
Bhubaneshwar  Info Stores               [High Value - At Risk]
Chennai       Surge Stores              [High Value - At Risk]
Hyderabad     Excel Stores                          [Champion]
Kochi         Expression                  [Low Value - Stable]
              Insight                   [Low Value - Churning]
              Surface Stores               [Mid Tier - Stable]
              Surge Stores              [High Value - At Risk]
Name: Segment, dtype: object


- This is the same pattern as Delhi NCR, just at smaller scale: Chennai's 42.8% decline and Bhubaneshwar's shaky numbers both trace to *High Value - At Risk* accounts, while Hyderabad's +14% growth is anchored by a "Champion" account. Bengaluru's drop to zero fits too — its 2018 revenue came from customers (like Electricalsara Stores, Excel Stores) who are primarily based in other markets and simply didn't make those incidental purchases again in 2019.
- Electricalsara Stores and Info Stores are the same accounts identified in the customer concentration analysis as part of the Top-10/Critical-tier group responsible for 86.9% of the total revenue decline. Their pullback isn't isolated to Delhi NCR, it shows up in Bengaluru's disappearance and Bhubaneshwar's instability too.
- Kochi (+5.5%, 4 customers, more diversified base) and Hyderabad (+14%, anchored by a Champion account) look like the genuine diversification opportunities in South, independent of the at-risk account cohort.

##### Conclusion - 03 identified a small set of named accounts driving the decline. Does that concentration also have a geographic signature and where - independent of those accounts, is genuine diversification actually happening?

The decline is driven by a specific cohort of accounts (the same Critical-tier accounts from the customer concentration analysis), and this is visible geographically across both North and South.

**What this confirms:**
- North's 21.5% decline and Delhi NCR's 22% are concentrated in "High Value - At Risk" accounts already named in the customer concentration analysis.
- The same accounts' pullback explains Chennai's decline, Bhubaneshwar's instability, and Bengaluru's disappearance, this isn't a North-specific or South-specific issue, it's an account-level issue with a geographic footprint.

**New finding:**
- Hyderabad and Kochi are concrete, named diversification targets — independent of the at-risk cohort and showing organic growth.

**Limitations:**
- South's market-level numbers are based on very small samples (Bhubaneshwar: 112 orders; most markets: 1 customer), so trends here should be read as directional, not statistically robust.